File: HavenApp/HavenApp/Services/PowPreferences.swift
import Foundation

@MainActor
class PowPreferences: ObservableObject {
    static let shared = PowPreferences()

    nonisolated static let minDifficulty = 8
    nonisolated static let maxDifficulty = 32

    private struct Keys {
        static let noteEnabled = "pow_note_enabled"
        static let noteDifficulty = "pow_note_difficulty"
        static let reactionEnabled = "pow_reaction_enabled"
        static let reactionDifficulty = "pow_reaction_difficulty"
        static let dmEnabled = "pow_dm_enabled"
        static let dmDifficulty = "pow_dm_difficulty"
    }

    @Published var notePowEnabled: Bool {
        didSet { UserDefaults.standard.set(notePowEnabled, forKey: Keys.noteEnabled) }
    }
    @Published var noteDifficulty: Int {
        didSet {
            let clamped = Self.clamp(noteDifficulty)
            if clamped != noteDifficulty { noteDifficulty = clamped; return }
            UserDefaults.standard.set(noteDifficulty, forKey: Keys.noteDifficulty)
        }
    }
    @Published var reactionPowEnabled: Bool {
        didSet { UserDefaults.standard.set(reactionPowEnabled, forKey: Keys.reactionEnabled) }
    }
    @Published var reactionDifficulty: Int {
        didSet {
            let clamped = Self.clamp(reactionDifficulty)
            if clamped != reactionDifficulty { reactionDifficulty = clamped; return }
            UserDefaults.standard.set(reactionDifficulty, forKey: Keys.reactionDifficulty)
        }
    }
    @Published var dmPowEnabled: Bool {
        didSet { UserDefaults.standard.set(dmPowEnabled, forKey: Keys.dmEnabled) }
    }
    @Published var dmDifficulty: Int {
        didSet {
            let clamped = Self.clamp(dmDifficulty)
            if clamped != dmDifficulty { dmDifficulty = clamped; return }
            UserDefaults.standard.set(dmDifficulty, forKey: Keys.dmDifficulty)
        }
    }

    private init() {
        let d = UserDefaults.standard
        self.notePowEnabled = d.object(forKey: Keys.noteEnabled) as? Bool ?? true
        self.noteDifficulty = Self.clamp(d.object(forKey: Keys.noteDifficulty) as? Int ?? 16)
        self.reactionPowEnabled = d.object(forKey: Keys.reactionEnabled) as? Bool ?? true
        self.reactionDifficulty = Self.clamp(d.object(forKey: Keys.reactionDifficulty) as? Int ?? 12)
        self.dmPowEnabled = d.object(forKey: Keys.dmEnabled) as? Bool ?? true
        self.dmDifficulty = Self.clamp(d.object(forKey: Keys.dmDifficulty) as? Int ?? 12)
    }

    nonisolated private static func clamp(_ n: Int) -> Int {
        max(minDifficulty, min(maxDifficulty, n))
    }

    /// Thread-safe snapshot for background use (reads UserDefaults directly).
    nonisolated static func snapshot() -> Snapshot {
        let d = UserDefaults.standard
        return Snapshot(
            noteEnabled: d.object(forKey: Keys.noteEnabled) as? Bool ?? true,
            noteDifficulty: clamp(d.object(forKey: Keys.noteDifficulty) as? Int ?? 16),
            reactionEnabled: d.object(forKey: Keys.reactionEnabled) as? Bool ?? true,
            reactionDifficulty: clamp(d.object(forKey: Keys.reactionDifficulty) as? Int ?? 12),
            dmEnabled: d.object(forKey: Keys.dmEnabled) as? Bool ?? true,
            dmDifficulty: clamp(d.object(forKey: Keys.dmDifficulty) as? Int ?? 12)
        )
    }

    struct Snapshot: Sendable {
        let noteEnabled: Bool
        let noteDifficulty: Int
        let reactionEnabled: Bool
        let reactionDifficulty: Int
        let dmEnabled: Bool
        let dmDifficulty: Int
    }
}

File: NostrVault/app/src/main/java/com/nostrvault/data/local/PowPreferences.kt
package com.nostrvault.data.local

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Proof-of-work preferences for NIP-13 mining.
 * Backed by SharedPreferences for immediate persistence.
 * Port of PowPreferences.swift.
 */
@Singleton
class PowPreferences @Inject constructor(
    @ApplicationContext context: Context,
) {
    companion object {
        const val MIN_DIFFICULTY = 8
        const val MAX_DIFFICULTY = 32

        private const val PREFS_NAME = "pow_preferences"
        private const val KEY_NOTE_ENABLED = "pow_note_enabled"
        private const val KEY_NOTE_DIFFICULTY = "pow_note_difficulty"
        private const val KEY_REACTION_ENABLED = "pow_reaction_enabled"
        private const val KEY_REACTION_DIFFICULTY = "pow_reaction_difficulty"
        private const val KEY_DM_ENABLED = "pow_dm_enabled"
        private const val KEY_DM_DIFFICULTY = "pow_dm_difficulty"
    }

    data class Snapshot(
        val noteEnabled: Boolean,
        val noteDifficulty: Int,
        val reactionEnabled: Boolean,
        val reactionDifficulty: Int,
        val dmEnabled: Boolean,
        val dmDifficulty: Int,
    )

    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private val _noteEnabled = MutableStateFlow(prefs.getBoolean(KEY_NOTE_ENABLED, false))
    val noteEnabled: StateFlow<Boolean> = _noteEnabled.asStateFlow()

    private val _noteDifficulty = MutableStateFlow(prefs.getInt(KEY_NOTE_DIFFICULTY, 16))
    val noteDifficulty: StateFlow<Int> = _noteDifficulty.asStateFlow()

    private val _reactionEnabled = MutableStateFlow(prefs.getBoolean(KEY_REACTION_ENABLED, false))
    val reactionEnabled: StateFlow<Boolean> = _reactionEnabled.asStateFlow()

    private val _reactionDifficulty = MutableStateFlow(prefs.getInt(KEY_REACTION_DIFFICULTY, 12))
    val reactionDifficulty: StateFlow<Int> = _reactionDifficulty.asStateFlow()

    private val _dmEnabled = MutableStateFlow(prefs.getBoolean(KEY_DM_ENABLED, false))
    val dmEnabled: StateFlow<Boolean> = _dmEnabled.asStateFlow()

    private val _dmDifficulty = MutableStateFlow(prefs.getInt(KEY_DM_DIFFICULTY, 12))
    val dmDifficulty: StateFlow<Int> = _dmDifficulty.asStateFlow()

    /** Thread-safe snapshot for background signing. */
    fun snapshot(): Snapshot = Snapshot(
        noteEnabled = prefs.getBoolean(KEY_NOTE_ENABLED, false),
        noteDifficulty = prefs.getInt(KEY_NOTE_DIFFICULTY, 16),
        reactionEnabled = prefs.getBoolean(KEY_REACTION_ENABLED, false),
        reactionDifficulty = prefs.getInt(KEY_REACTION_DIFFICULTY, 12),
        dmEnabled = prefs.getBoolean(KEY_DM_ENABLED, false),
        dmDifficulty = prefs.getInt(KEY_DM_DIFFICULTY, 12),
    )

    fun setNoteEnabled(enabled: Boolean) {
        _noteEnabled.value = enabled
        prefs.edit().putBoolean(KEY_NOTE_ENABLED, enabled).apply()
    }

    fun setNoteDifficulty(difficulty: Int) {
        val clamped = difficulty.coerceIn(MIN_DIFFICULTY, MAX_DIFFICULTY)
        _noteDifficulty.value = clamped
        prefs.edit().putInt(KEY_NOTE_DIFFICULTY, clamped).apply()
    }

    fun setReactionEnabled(enabled: Boolean) {
        _reactionEnabled.value = enabled
        prefs.edit().putBoolean(KEY_REACTION_ENABLED, enabled).apply()
    }

    fun setReactionDifficulty(difficulty: Int) {
        val clamped = difficulty.coerceIn(MIN_DIFFICULTY, MAX_DIFFICULTY)
        _reactionDifficulty.value = clamped
        prefs.edit().putInt(KEY_REACTION_DIFFICULTY, clamped).apply()
    }

    fun setDmEnabled(enabled: Boolean) {
        _dmEnabled.value = enabled
        prefs.edit().putBoolean(KEY_DM_ENABLED, enabled).apply()
    }

    fun setDmDifficulty(difficulty: Int) {
        val clamped = difficulty.coerceIn(MIN_DIFFICULTY, MAX_DIFFICULTY)
        _dmDifficulty.value = clamped
        prefs.edit().putInt(KEY_DM_DIFFICULTY, clamped).apply()
    }

    /** Resolve difficulty for a given event kind. Returns 0 if PoW is disabled for that kind. */
    fun difficultyForKind(kind: Int): Int {
        val snap = snapshot()
        return when (kind) {
            1, 6, 30023 -> if (snap.noteEnabled) snap.noteDifficulty else 0
            7 -> if (snap.reactionEnabled) snap.reactionDifficulty else 0
            4, 14 -> if (snap.dmEnabled) snap.dmDifficulty else 0
            else -> 0
        }
    }
}
